[DWB] Option to limit velocity commands in trajectory generator

huiyulhy opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Issue #3277 |
| Primary OS tested on | Ubuntu 22.04|
| Robotic platform tested on | Turtlebot3 Gazebo simulation - waffle |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points
- Added an option in the standard trajectory generator in DWB to allow the user to limit the velocity command in the trajectory 
-based on current trajectory through a new parameter (`limit_vel_cmd_in_traj_`)
- This option ensures that the acceleration constraints are respected in the event there is a large difference between the current and commanded velocity - which may be more useful when there is no velocity smoother in use
- `limit_vel_cmd_in_traj_` is set to false by default, but when set to true, would use the first velocity computed in the trajectory rollout as the trajectory velocity

Default behavior: 
![dev_default](https://github.com/user-attachments/assets/0259c2b5-06b2-42ee-86d6-eb73da14de23)

When `limit_vel_cmd_in_traj_` is set to `true`:
![proposed_default](https://github.com/user-attachments/assets/4313e983-416c-48eb-94e0-c4470fc513a4)

We observed slightly improved velocity tracking, and similar time to reach the same goal setting this parameter to true.

## Description of documentation updates required from your changes
- Added parameter `limit_vel_cmd_in_traj_` to DWB local planner (in standard trajectory generator)
---

## Future work that may be required in bullet points
- This parameter could also be extensible to other trajectory generation methods within DWB

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:
```diff
@@ -75,6 +75,10 @@ void StandardTrajectoryGenerator::initialize(
 
+  nav2_util::declare_parameter_if_not_declared(
+    nh,
+    plugin_name + ".limit_vel_cmd_in_traj_", rclcpp::ParameterValue(false));
```
```suggestion
    plugin_name + ".limit_vel_cmd_in_traj", rclcpp::ParameterValue(false));
```

huiyulhy replied:
Updated this, thanks for catching!

SteveMacenski commented on nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:
```diff
@@ -89,6 +93,7 @@ void StandardTrajectoryGenerator::initialize(
   nh->get_parameter(plugin_name + ".linear_granularity", linear_granularity_);
   nh->get_parameter(plugin_name + ".angular_granularity", angular_granularity_);
   nh->get_parameter(plugin_name + ".include_last_point", include_last_point_);
+  nh->get_parameter(plugin_name + ".limit_vel_cmd_in_traj_", limit_vel_cmd_in_traj_);
```
```suggestion
  nh->get_parameter(plugin_name + ".limit_vel_cmd_in_traj", limit_vel_cmd_in_traj_);
```

huiyulhy replied:
Updated too

SteveMacenski commented on nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:
```diff
@@ -156,9 +161,14 @@ dwb_msgs::msg::Trajectory2D StandardTrajectoryGenerator::generateTrajectory(
   double running_time = 0.0;
   std::vector<double> steps = getTimeSteps(cmd_vel);
   traj.poses.push_back(start_pose);
+  bool firstVel = false;
```
Variables should be `snake_case` not `camelCase`

huiyulhy replied:
Updated this, thanks for pointing it out!

SteveMacenski reviewed (changes requested):
Great! We also need:
- Nav2 migration guide updated to include this change and the new parameter
- Nav2 configuration guide for DWB including this new parameter
- You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

huiyulhy commented:
> Great! We also need:
> 
> * Nav2 migration guide updated to include this change and the new parameter
> * Nav2 configuration guide for DWB including this new parameter
> * You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

Thanks for the suggestion! Am updating for the documentation for Nav2 and I wanted to confirm that the addition of this change in the migration guide would be from Jazzy to K-Turtle?

SteveMacenski commented:
That is correct!

huiyulhy commented:
> Great! We also need:
> 
> * Nav2 migration guide updated to include this change and the new parameter
> * Nav2 configuration guide for DWB including this new parameter
> * You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

@SteveMacenski thanks for the feedback! Fixed the linting errors and updated the documentation in [this PR](https://github.com/ros-navigation/docs.nav2.org/pull/591) on nav2 docs!
